File: src/main/kotlin/ru/herobrine1st/ocbridge/network/Service.kt
package ru.herobrine1st.ocbridge.network

import com.google.gson.Gson
import com.google.gson.JsonArray
import ru.herobrine1st.ocbridge.OCBridge
import ru.herobrine1st.ocbridge.data.PingRequest
import ru.herobrine1st.ocbridge.data.RequestStructure
import ru.herobrine1st.ocbridge.integration.RequestBuilder
import ru.herobrine1st.ocbridge.integration.Response
import java.nio.ByteBuffer
import java.nio.channels.SocketChannel
import kotlin.math.pow
import kotlin.random.Random

abstract class Service(val name: String, val password: String) {
    var channel: SocketChannel? = null
        set(value) {
            field = value
            pending.clear()
        }
    val pending = ArrayList<RequestStructure>()
    val callbacks = HashMap<String, (Response) -> Unit>()
    var pendingRemove: Boolean = false
    private var lastPingTimestamp = 0L
    val isReady
        get() = channel != null
    val isNotReady
        get() = channel == null

    /**
     * On connect client hook.
     */
    abstract fun onConnect()

    /**
     * On disconnect client hook.
     */
    abstract fun onDisconnect()

    /**
     * On message hook. Called when client passed a message to the bridge.
     */
    abstract fun onMessage(message: String)

    /**
     * On event hook. Called when client has OpenComputers event(s).
     * Events will dispatched one by one even if client sent more than one event.
     */
    abstract fun onEvent(event: JsonArray)

    fun disconnect() {
        channel?.close()
        channel = null
        onDisconnect()
    }

    /**
     * Create a request builder
     */
    fun request(): RequestBuilder {
        var hash = Random.nextLong()
        while(pending.any { it.hash == hash.toString() }) hash = Random.nextLong()
        return RequestBuilder(hash, this)
    }

    fun executeRequest(structure: RequestStructure, callback: (Response) -> Unit) {
        if(!isReady) throw IllegalStateException()
        channel!!.write(ByteBuffer.wrap("${Gson().toJson(structure)}\n".toByteArray()))
        structure.timestamp = System.nanoTime()
        callbacks[structure.hash] = callback
        pending += structure
    }

    open fun heartbeat() {
        if(!isReady) return
        if(pending.any { it.type == RequestStructure.Type.PING }
            && System.nanoTime() - lastPingTimestamp > 5 * 10.0.pow(9.0)) {
                SocketThread.logger.info("$name disconnected cause of no response")
                disconnect()
        } else if (System.nanoTime() - lastPingTimestamp > 5 * 10.0.pow(9.0)) {
            var hash = Random.nextLong()
            while(pending.any { it.hash == hash.toString() }) hash = Random.nextLong()
            val req = PingRequest(hash)
            pending.add(req)
            channel?.writeJson(req)
            lastPingTimestamp = req.timestamp
        }
    }
}

File: src/main/python/ocbridge/service.py
import abc
import logging

from ocbridge.utils import model_to_bytes
from socket import socket
from typing import Optional
from ocbridge.structure import *
logger = logging.getLogger("OCBridge")

class Service(abc.ABC):
    connection: Optional[socket]
    name: str
    password: str
    pending_remove: bool
    pending: list
    callbacks: dict
    last_ping_timestamp: int

    def __init__(self, name: str, password: str):
        self.connection = None
        self.name = name
        self.password = password
        self.last_ping_timestamp = 0

    @property
    def is_ready(self):
        return self.connection is not None

    def connect(self, connection):
        self.connection = connection
        self.on_connect()

    def disconnect(self):
        if self.connection:
            self.connection.close()
        self.connection = None
        self.on_disconnect()

    @abc.abstractmethod
    def on_connect(self):
        pass

    @abc.abstractmethod
    def on_disconnect(self):
        pass

    @abc.abstractmethod
    def on_message(self, message: str):
        pass

    @abc.abstractmethod
    def on_event(self, event: tuple):
        pass

    def heartbeat(self):
        if not self.is_ready: return
        if time.time() - self.last_ping_timestamp > 5:
            if next(filter(lambda x: x.type == RequestStructure.Type.PING,self.pending), default=False):
                self.disconnect()
                logger.info(f"{self.name} disconnected cause of no response")
        else:
            req = PingStructure()
            self.pending.append(req)
            self.connection.send(model_to_bytes(req))
            self.last_ping_timestamp = req.timestamp